### PR TITLE
Unit tests for MatchingEngine component

### PR DESCRIPTION
**Pull request by lmpn**
Added unit tests for the API these tests are divided in expiration, queries and placing/canceling orders.

### CODE REVIEW

**Comment by teixeiramarcia** [comment]
+1

**Comment by teixeiramarcia** [comment]
👍

File: apps/exchange/test/matching_engine_test.exs [hunk 1]
```diff
@@ -0,0 +1,569 @@
...
+    end
+
+    test "orders fullfilled are not added to expiration_list", %{order_book: ob} do
+      t1 = :os.system_time(:millisecond)
+
+      buy_order =
+        sample_expiring_order(%{size: 750, price: 4010, side: :buy, id: "9", exp_time: t1})
+
+      new_book = OrderBook.price_time_match(ob, buy_order)
+      assert [] == new_book.expiration_list
```
Comment (Qu4tro) [inline apps/exchange/test/matching_engine_test.exs:109]: https://en.wikipedia.org/wiki/Yoda_conditions

File: apps/exchange/test/matching_engine_test.exs [hunk 2]
```diff
@@ -0,0 +1,569 @@
...
+
+      sell_order =
+        sample_expiring_order(%{size: 1000, price: 4020, side: :sell, id: "10", exp_time: t2})
+
+      {:reply, _code, ob} = MatchingEngine.handle_call({:place_limit_order, buy_order}, nil, ob)
+      {:reply, _code, ob} = MatchingEngine.handle_call({:place_limit_order, sell_order}, nil, ob)
+      order_id_1 = buy_order.order_id
+      order_id_2 = sell_order.order_id
+
+      assert [{t2, order_id_2}, {t1, order_id_1}] == ob.expiration_list
```
Comment (Qu4tro) [inline apps/exchange/test/matching_engine_test.exs:99]: Yoda condition?

File: apps/exchange/test/matching_engine_test.exs [hunk 3]
```diff
@@ -0,0 +1,569 @@
...
+      order = sample_order(%{size: 100, price: 10_000, side: :sell})
+      order = %Order{order | type: :market, order_id: "4"}
+      {:reply, code, _ob} = MatchingEngine.handle_call({:place_market_order, order}, nil, order_book)
+      assert code == :error
+    end
+
+    test "Cancel existing order", %{order_book: order_book} do
+      {:reply, code, _ob} = MatchingEngine.handle_call({:cancel_order, "4"}, nil, order_book)
+      assert code == :ok
+    end
```
Comment (Qu4tro) [inline apps/exchange/test/matching_engine_test.exs:272]: Este teste, não devia verificar que a ordem teria sido cancelada com sucesso (para além de verificar o code == :ok ).

Comment (lmpn) [reply apps/exchange/test/matching_engine_test.exs:422]: done

File: apps/exchange/test/matching_engine_test.exs (hunk 2 above)
Comment (lmpn) [reply apps/exchange/test/matching_engine_test.exs]: Refactored

File: apps/exchange/test/matching_engine_test.exs (hunk 1 above)
Comment (lmpn) [reply apps/exchange/test/matching_engine_test.exs]: Refactored

File: apps/exchange/test/order_book_test.exs [hunk 4]
```diff
@@ -485,6 +485,15 @@ defmodule OrderBookTest do
       assert OrderBook.total_ask_orders(order_book) == 4
     end
 
+    test "All open orders in default order_book", %{order_book: order_book} do
+      assert Enum.count(OrderBook.open_orders(order_book)) == 8
+    end
+
+    test "All open orders by trader_id in default order_book", %{order_book: order_book} do
+      assert Enum.count(OrderBook.open_orders_by_trader(order_book, "alchemist")) == 8
+      assert OrderBook.open_orders_by_trader(order_book, "foo") == []
```
[the remark this answers is not in the document]
Comment (lmpn) [reply apps/exchange/test/order_book_test.exs:495]: O mix credo dá warnings se não estiver assim

File: apps/exchange/lib/exchange/order_book.ex [hunk 5]
```diff
@@ -367,10 +371,13 @@ defmodule Exchange.OrderBook do
         !Enum.empty?(Map.get(order_book.sell, pp))
       end)
       |> List.first()
-      new_ask_min = case new_ask_min do
+
+    new_ask_min =
+      case new_ask_min do
         nil -> 99_999
```
Comment (emanuel) [inline apps/exchange/lib/exchange/order_book.ex:377]: I think  that now ask_min can't be nil ...  but we have to make sure...

Comment (lmpn) [reply apps/exchange/lib/exchange/order_book.ex:377]: Yes it can and happens when the the sell side is empty

**Pull request merged**